### PR TITLE
Add documentation to sync script and read the docs configuration

### PR DESCRIPTION
**Pull request by grantseltzer**
The goal of this PR is to:

1) Sync the documentation from the kernel repo (in `Documentation/bpf/libbpf`) to the mirror
2) Add configuration so that readthedocs.io can pull in the documentation files that are synced, plus the API autogenerated docs.

@anakryiko I'm not 100% confident that the added line in `scripts/sync-kernel.sh` is sufficient to meet goal #1. I believe this will only find changes going forward. I tried running it with the base commit that I added the documentation files to the kernel [1]. However that doesn't work since it tracks changes since then. If I set the base commit the commit before [1] it fails because the path `Documentation/bpf/libbpf` does not yet exist. This may require manually committing the files here on github? I may misunderstand the sync script, I'm a novice at bash.

To accomplish our goals of having a robust libbpf documentation site, two more things have to be done in addition to this PR:

1) Point libbpf.readthedocs.org at this repo (Andrii, you and I both have admin privileges, this is as simple as changing a setting in the web UI)

2) A couple of changes to the kernel repo documentation files:
    - Removing the API document from the kernel repo. I believe that this doesn't belong in the kernel repository mostly because it won't be properly versioned.
    - Remove the 'TOC' (table of contents) from `Documentation/bpf/index.rst` which points to the libbpf documents individually, and instead have an index file in `Documentation/bpf/libbpf` so that the mirror syncs seamlessly.

I submitted a patch handling the second point above [here](https://lore.kernel.org/bpf/20210810020508.280639-1-grantseltzer@gmail.com/T/#u), and will handle the first one once this PR is merged.

[1] `f42cfb469f9b4a1c002a03cce3d9329376800a6f` 

Signed-off-by: grantseltzer <grantseltzer@gmail.com>

### CODE REVIEW

File: scripts/sync-kernel.sh [hunk 1]
```diff
@@ -47,6 +47,7 @@ PATH_MAP=(									\
 	[tools/include/uapi/linux/netlink.h]=include/uapi/linux/netlink.h	\
 	[tools/include/uapi/linux/pkt_cls.h]=include/uapi/linux/pkt_cls.h	\
 	[tools/include/uapi/linux/pkt_sched.h]=include/uapi/linux/pkt_sched.h	\
+	[Documentation/bpf/libbpf]=src/docs	\
```
Comment (anakryiko) [inline scripts/sync-kernel.sh:50]: nit: please keep `\` aligned

File: scripts/sync-kernel.sh [hunk 2]
```diff
@@ -47,6 +47,7 @@ PATH_MAP=(									\
...
 )
 
 LIBBPF_PATHS="${!PATH_MAP[@]} :^tools/lib/bpf/Makefile :^tools/lib/bpf/Build :^tools/lib/bpf/.gitignore :^tools/include/tools/libc_compat.h"
```
Comment (anakryiko) [inline scripts/sync-kernel.sh:53]: I'm currently getting this when running sync script:

```
...
Comparing list of files...
--- /tmp/tmp.0XYbOenSZb/linux-view.ls   2021-08-11 12:03:46.248692923 -0700
+++ /tmp/tmp.0XYbOenSZb/github-view.ls  2021-08-11 12:03:46.250692920 -0700
@@ -17,10 +17,12 @@
 src/btf.c
 src/btf.h
 src/btf_dump.c
-src/docs/libbpf.rst
-src/docs/libbpf_api.rst
-src/docs/libbpf_build.rst
-src/docs/libbpf_naming_convention.rst
+src/docs/.gitignore
+src/docs/api.rst
+src/docs/conf.py
+src/docs/sphinx/Makefile
+src/docs/sphinx/doxygen/Doxyfile
+src/docs/sphinx/requirements.txt
 src/gen_loader.c
 src/hashmap.c
 src/hashmap.h
```

Please adjust LIBBPF_VIEW_EXCLUDE_REGEX and LINUX_VIEW_EXCLUDE_REGEX accordingly to make any differences go away.

**Review by anakryiko — CHANGES_REQUESTED** [review]
We might have discussed this already, sorry, but why are we putting `docs` under `src`? Docs are docs, source code is source code. Can we have `/src` and `/docs` at the same level?

Also small nit. You have:

```
Signed-off-by: grantseltzer <grantseltzer@gmail.com>
```

Please make sure you spell your full name correctly in Signed-off-by.

**Comment by anakryiko** [comment]
Also a question about your changes in https://patchwork.kernel.org/project/netdevbpf/patch/20210810020508.280639-1-grantseltzer@gmail.com/

Why are you removing links to libbpf/libbpf, libbpf/libbpf_build, libbpf/libbpf_naming_convention, not just libbpf/libbpf_api?

Also:
```
Documentation/bpf/libbpf/libbpf.rst is a userspace library for loading and interacting with bpf programs.
```

That path to .rst file, is that going to be automatically converted to a link or something? How can I try to see those changes locally before pushing them to bpf-next/master? Sorry for stupid questions.

**Comment by grantseltzer** [comment]
> Also a question about your changes in https://patchwork.kernel.org/project/netdevbpf/patch/20210810020508.280639-1-grantseltzer@gmail.com/
> 
> Why are you removing links to libbpf/libbpf, libbpf/libbpf_build, libbpf/libbpf_naming_convention, not just libbpf/libbpf_api?

This is to make the mirror syncing seamless. The paths would be different in the mirror and therefore we'd have to rewrite the links in the sync script, or something of that nature. The readthedocs UI automatically creates a TOC on the left side.

> 
> Also:
> 
> ```
> Documentation/bpf/libbpf/libbpf.rst is a userspace library for loading and interacting with bpf programs.
> ```
> 
> That path to .rst file, is that going to be automatically converted to a link or something? How can I try to see those changes locally before pushing them to bpf-next/master? Sorry for stupid questions.

I must insist there's no such thing as stupid questions. 

Yes it will be converted to a link. You can run `make htmldocs` to see this.

File: scripts/sync-kernel.sh (hunk 2 above)
Comment (grantseltzer) [reply scripts/sync-kernel.sh:53]: I can't reproduce this issue running it how you told me to, I'm not sure what the output is saying, what causes this?

**Comment by anakryiko** [comment]
There were few small issues with the sync script. And when starting to sync some new files which are already in kernel tree, the initial version of those files have to be copied to libbpf as is in a separate commit. Please see https://github.com/libbpf/libbpf/pull/361 for all the changes that I had to do to make sync-script work with docs.

It took quite a few trials to figure all this out even for me (and I sort of knew what I'm looking for), so I figured it's not fair to ask you to figure this out on your own.

**Pull request merged**